elasticdl/python/worker/allreduce_trainer.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2020 The ElasticDL Authors. All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import time
+
+import tensorflow as tf
+from tensorflow.python.framework.errors_impl import UnknownError
+
+from elasticdl.python.common.constants import HorovodEnv
+from elasticdl.python.common.log_utils import default_logger as logger
+
+try:
+    import horovod.tensorflow as hvd
+except ImportError:
+    hvd = None
+
+
+# The default maximum number of retries for allreduce operation
+# if allreduce-based distributed training strategy is used.
+DEFAULT_MAX_ALLREDUCE_RETRY_NUM = 5
+
+
+class AllReduceTrainer(object):
```

Comment:
Will we have a ParameterServerTrainer? How about move some common method into an abstract class?
These two classes will implement these methods.

Reply:
Yes, we can separate training codes from worker.py. I have created #2253 to develop an abstract class and ParameterServerTrainer.